src/util.cc

```diff
@@ -825,6 +825,36 @@ double GetLoadAverage() {
 }
 #else
 double GetLoadAverage() {
+# if defined(__linux__)
+  // Only try to open the file once.
+  static int fd = -1;
+  if (fd == -1) {
+    fd = open("/proc/loadavg", O_RDONLY);
+    if (fd < 0) {
+      fd = -2;
+    }
+  }
+
+  if (fd >= 0) {
+    char buf[65], *p;
+    ssize_t nbr = pread(fd, buf, sizeof(buf) - 1, 0);
+    // Load average should be at least "1 5 9 n/".
+    if (nbr < 8) {
+      return -0.0f;
+    }
+    p = strchr(buf, ' ');
+    if (p) {
+      p = strchr(p + 1, ' ');
+    }
+    if (p) {
+      p = strchr(p + 1, ' ');
+    }
+    if (p && (unsigned(p[1]) - '0' <= 9)) {
```

Comment:
how do you know `p+1 != NULL`?

Reply:
Would it be sufficient to set buf[nbr] = '\0' after checking that nbr is large enough?  p+1 == NULL only if the kernel gives us a corrupted buffer from pread() and the stack happened to be near the top of the address stack and nothing higher on the stack was either '\0' or ' '.

But if the kernel gives us a corrupted buffer, we could read past the end of buf[], so it is safer to explicitly stick a '\0' at the end.  Then we know that at this point, (p == NULL) || (p[0] == ' ' and p[0] has at least '\0' afterwards).

Reply:
Updated in 367fbf8d8.